### PR TITLE
BE: IQueryableExtensions.OrderBy now is working with properties of derived types.

### PR DESCRIPTION
**Pull request by blazemachine**
Effective type (type of elements in list) was added to code.

### CODE REVIEW

File: VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs [hunk 1]
```diff
@@ -69,39 +69,40 @@ public static IOrderedQueryable<T> ThenBySortInfos<T>(this IOrderedQueryable<T>
...
-				throw new ArgumentNullException("property");
+            if (property == null)
+                throw new ArgumentNullException("property");
 
-			string[] props = property.Split('.');
-			Type type = typeof(T);
-			ParameterExpression arg = Expression.Parameter(type, "x");
-			Expression expr = arg;
+            string[] props = property.Split('.');
+            var effectiveType = source.FirstOrDefault().GetType();
```
Comment (artem-dudarev) [inline VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs]: FirstOrDefault() will lead to redundant and unexpected request to the database.

Comment (yecli) [reply VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs]: Really, it is non-acceptable solution. Cannot find a way to get specific element runtime type besides getting element. [IQueryable.ElementType](https://docs.microsoft.com/en-US/dotnet/api/system.linq.iqueryable.elementtype?view=netframework-4.7.2) does not help.  [System.Linq.Dynamic](https://github.com/kahanu/System.Linq.Dynamic) cannot handle such situations too.
So we should admit that we cannot implement sorting using derived class new properties in this extension without immediate execution.
The only viable option I see is passing type as propertyName, in string. 
So need to decide if additional extension like .OrderBy(string propertyName, string elementTypeName) could be useful for us.

Comment (yecli) [reply VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs]: Decided to use type registered in AbstractTypeFactory for base type.

File: VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs [hunk 2]
```diff
@@ -68,40 +67,42 @@ public static IOrderedQueryable<T> ThenBySortInfos<T>(this IOrderedQueryable<T>
...
         public static IOrderedQueryable<T> ApplyOrder<T>(IQueryable<T> source, string property, string methodName)
-		{
-			if (property == null)
-				throw new ArgumentNullException("property");
+        {
+            if (property == null)
+                throw new ArgumentNullException(nameof(property));
+
+            var props = property.Split('.');
+            var registeredType = AbstractTypeFactory<T>.AllTypeInfos.Select(x => x.Type).FirstOrDefault();
```
Comment (yecli) [inline VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs]: Better to use AbstractTypeFactory<T>.TryCreateInstance(), as it consider inheritance.

File: VirtoCommerce.Platform.Tests/Common/QueryableExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using VirtoCommerce.Platform.Core.Common;
+using Xunit;
+
+namespace VirtoCommerce.Platform.Tests.Common
+{
+    [CLSCompliant(false)]
+    public class QueryableExtensionsTests
```
Comment (yecli) [inline VirtoCommerce.Platform.Tests/Common/QueryableExtensionsTests.cs]: Please add     
[Trait("Category", "CI")] 
to include these tests into CI process

Comment (anivirtoway) [reply VirtoCommerce.Platform.Tests/Common/QueryableExtensionsTests.cs]: fix

File: VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs (hunk 2 above)
Comment (anivirtoway) [reply VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs]: fix

File: VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs (hunk 1 above)
Comment (yecli) [reply VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs]: @artem-dudarev Could you check new implementation?

**Comment by anivirtoway** [comment]
To use EF you cannot use the expression "Expression expr = Expression.Convert (arg, effectiveType);" more details can be found here https://entityframework.net/knowledge-base/47425013/linq-to-entities-only-supports-casting-edm-primitive-or-enumeration-types--unique-new-post-
Unfortunately I did not find an alternative

**Comment by tatarincev** [comment]
Have been moved to On Hold  because the proposed solution should not allow cases with sorting collections containing objects of different types.

**Pull request closed without merging**